staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go

```diff
@@ -69,50 +70,69 @@ func DefaultStacktracePred(status int) bool {
 	return (status < http.StatusOK || status >= http.StatusInternalServerError) && status != http.StatusSwitchingProtocols
 }
 
-// NewLogged turns a normal response writer into a logged response writer.
+// WithLogging wraps the handler with logging.
+func WithLogging(handler http.HandlerFunc, pred StacktracePred) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+		if _, err := respLoggerFromContext(req); err == nil {
+			panic("multiple NewLogging calls!")
+		}
+		rl := newLogged(req, w).StacktraceWhen(pred)
+		req = req.WithContext(context.WithValue(ctx, "httplog.respLogger", rl))
+
+		defer rl.Log()
+		handler(rl, req)
+	})
+}
+
+func respLoggerFromContext(req *http.Request) (*respLogger, error) {
+	ctx := req.Context()
+	val := ctx.Value("httplog.respLogger")
```

Comment:
We should use a unique int type for the key to avoid collisions.

Reply:
done